infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java

```diff
@@ -0,0 +1,73 @@
+package org.corfudb.infrastructure.logreplication.infrastructure;
+
+import org.corfudb.infrastructure.logreplication.infrastructure.plugins.ILogReplicationVersionAdapter;
+import org.corfudb.runtime.collections.TxnContext;
+
+/**
+ * Rolling upgrade handling means cluster must function in a mode where not all
+ * nodes are running the same codebase.
+ * Here the newer code base must function in a backward compatible mode until all
+ * the nodes have been upgraded to the same version.
+ * Once all nodes have upgraded it may atomically migrate data, clean up old code and switch
+ * completely to newer logic.
+ *
+ * To facilitate this in LR we add this module which would run once on startup as shown in the
+ * following diagram.
+ * In addition to running on startup, if cluster is in a mixed mode, all the callers who are
+ * mutating data in a new format need to check if upgrade is on and mutate data in a backward
+ * compatible manner.
+ *
+ * For anything post corfu-0.4.0.1 we consider the code base to be at "V2"
+ * Anything prior to and including corfu-0.4.0.1 is V1
+ * Any code shipped that uses Source/Sink can be thought of as "V2"
+ *                  ┌───────────────────────────────────┐
+ *                  │ isClusterAllAtV2 = false (in-mem) │
+ *                  └──────────────────┬────────────────┘
+ *                                     │
+ *                     ┌───────────────▼───────────┐
+ *                   ┌─┤                           ├─┐
+ *         Yes ┌─────┴─┴┐  isRollingUpgradeOn() ? ┌┴─┴──────┐ No
+ *             │        └─────────────────────────┘         │
+ *             │                                            │
+ * ┌───────────▼──────────────┐                         ┌──────▼──────────────────┐
+ * │ isClusterAllAtV2 = false │                         │ isClusterAllAtV2 = true │
+ * └───────────┬──────────────┘                         └───────┬─────────────────┘
+ *             │            ┌──────────────────────┐            │
+ *             └───────────►│ start Log Replicator │◄───────────┘
```

Comment:
I did not add it in this PR because we don't have any migration code yet, let me see if the startup code can still be modified to match above diagram.

Reply:
Done, added the check to LR startup.